site/js/modules/webui/input.js

```diff
@@ -0,0 +1,187 @@
+import * as Autograder from '../autograder/base.js';
+
+// A general representation of a user input field.
+class Field {
+    constructor(
+            name, displayName,
+            {
+                type = 'text', underlyingType = 'string', required = false, placeholder = '',
+                inputClass = '', attributes = '', labelBefore = true, extractInputFunc = undefined
+            } = {}) {
+        // The name of the field which will be used for targeting the field.
+        this.name = name;
+
+        // The display name that will be shown to the user.
+        this.displayName = displayName;
+
+        // The HTML input type.
+        // Supports checkbox, email, number, password, and text types.
+        this.type = type;
+
+        // The underlying type of the field.
+        // Non-standard types are parsed to JSON.
+        // If a non-standard type cannot be parsed to JSON,
+        // a validation error is raised.
+        this.underlyingType = underlyingType;
+
+        // Flags the field requires user input.
+        this.required = required;
+
+        // The placeholder text for the input.
+        this.placeholder = placeholder;
+
+        // Optional classes that are attached to the input.
+        this.inputClass = inputClass;
+
+        // Any additional attributes to the input field.
+        // If the field is required, the required attribute will be added automatically.
+        this.attributes = attributes;
```

Comment:
The naming here should probably be symmetric with inputClasses.
Maybe add an "extra" or "additional" prefix.